checker/jdk/index/src/java/nio/CharBuffer.java

```diff
@@ -0,0 +1,1523 @@
+     * to this buffer's content will be visible in the new buffer, and vice
+     * versa; the two buffers' position, limit, and mark values will be
+     * independent.
+     *
+     * <p> The new buffer's capacity, limit, position, and mark values will be
+     * identical to those of this buffer.  The new buffer will be direct if,
+     * and only if, this buffer is direct, and it will be read-only if, and
+     * only if, this buffer is read-only.  </p>
+     *
+     * @return  The new char buffer
+     */
+    public abstract CharBuffer duplicate();
+
+    /**
+     * Creates a new, read-only char buffer that shares this buffer's
+     * content.
+     *
+     * <p> The content of the new buffer will be that of this buffer.  Changes
+     * to this buffer's content will be visible in the new buffer; the new
+     * buffer itself, however, will be read-only and will not allow the shared
+     * content to be modified.  The two buffers' position, limit, and mark
+     * values will be independent.
+     *
+     * <p> The new buffer's capacity, limit, position, and mark values will be
+     * identical to those of this buffer.
+     *
+     * <p> If this buffer is itself read-only then this method behaves in
+     * exactly the same way as the {@link #duplicate duplicate} method.  </p>
+     *
+     * @return  The new, read-only char buffer
+     */
+    public abstract CharBuffer asReadOnlyBuffer();
+
+
+    // -- Singleton get/put methods --
+
+    /**
+     * Relative <i>get</i> method.  Reads the char at this buffer's
+     * current position, and then increments the position.
+     *
+     * @return  The char at the buffer's current position
+     *
+     * @throws  BufferUnderflowException
+     *          If the buffer's current position is not smaller than its limit
+     */
+    public abstract char get();
+
+    /**
+     * Relative <i>put</i> method&nbsp;&nbsp;<i>(optional operation)</i>.
+     *
+     * <p> Writes the given char into this buffer at the current
+     * position, and then increments the position. </p>
+     *
+     * @param  c
+     *         The char to be written
+     *
+     * @return  This buffer
+     *
+     * @throws  BufferOverflowException
+     *          If this buffer's current position is not smaller than its limit
+     *
+     * @throws  ReadOnlyBufferException
+     *          If this buffer is read-only
+     */
+    public abstract CharBuffer put(char c);
+
+    /**
+     * Absolute <i>get</i> method.  Reads the char at the given
+     * index.
+     *
+     * @param  index
+     *         The index from which the char will be read
+     *
+     * @return  The char at the given index
+     *
+     * @throws  IndexOutOfBoundsException
+     *          If <tt>index</tt> is negative
+     *          or not smaller than the buffer's limit
+     */
+    public abstract char get(@NonNegative @LessThan("this.limit") int index);
```

Comment:
The use of `this.limit` here smells odd to me. I would expect an `@IndexFor` annotation instead. It will be difficult for clients to establish an invariant against an internal property of the buffer. Perhaps we should change this to `@LessThan("this.limit()")`? Unless there is an appropriate `@IndexFor` annotation. What is the relationship between a buffer's length and its limit?